### PR TITLE
Replace ColdSignal and HotSignal with SignalProducer and Signal

### PR DESCRIPTION
**Pull request by jspahrsummers**
After the long discussion in #1650, I invited @NachoSoto, @andymatuschak, and @kastiglione to talk more in-depth about the Swift APIs, and specifically where they need improvement.

We had a really, really great brainstorming session, and this is the result. The good news is that we all feel pretty good about this direction, and the bad news is that everything is getting rewritten again. :trollface:

The important bits are documented on the types themselves, but here’s the handy Quick Start guide of the proposed changes:

1. `HotSignal<T>` will be replaced by `Signal<T>`. Besides the name change, these signals can now send `Error` and `Completed` events. This avoids a lot of the lifecycle weirdness that has come up with a value-only `HotSignal`.
1. `ColdSignal<T>` will be replaced by `SignalProducer<T>`. As the name implies, a signal producer creates `Signal`s (via the `start` method). You can kind of think of the relationship as something like:
 * `SignalProducer : Signal :: RACSignal : subscription`
 * `SignalProducer : Signal :: class : instance`
1. Most of the basic operators are written in terms of `Signal`, but can be lifted to apply to `SignalProducer` instead.
1. Operators that benefit from “cold” semantics, like `retry` and `repeat`, are written in terms of `SignalProducer`.

The advantages here are:

* The relationship between “hot” and “cold” signals is now much more explicit and well-defined. A “cold” signal (a producer) creates “hot” signals, each of which represents a “subscription” (in RAC 2.x terms).
* Many operators only have to be written once, in terms of `Signal`, and can be lifted to the `SignalProducer` level.
* The semantics and/or difficulties of converting between “hot” and “cold” are more obvious, I think. You don’t need to swap between them on the fly, because they mean different things.
* Signal lifetime is now consistent, and completely defined by the event stream. Signal producers don’t do anything special—they really are just value types, and don’t encode any information about lifetime.

I’ll also make some notes in the changes below, to explain more specific things. Please leave feedback if you have it! :bow:

### CODE REVIEW

**Comment by joshaber** [comment]
:sparkling_heart: This looks fantastic :sparkling_heart: 

A single signal type is awesome. I don't love the name of `SignalProducer` but I don't have any better ideas that haven't already been discussed.

--

Do we want to go ahead and merge this?

**Comment by robb** [comment]
LGTM

**Comment by jspahrsummers** [comment]
> Do we want to go ahead and merge this?

I'll go ahead and build the implementations here, unless there's a particular reason not to do that.

**Comment by joshaber** [comment]
:+1:

**Comment by kastiglione** [comment]
> I don't love the name of `SignalProducer` but I don't have any better ideas that haven't already been discussed.

It's a shame english cannot handle our scale.

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+	}
+}
+
+/// Combines the latest value of the receiver with the latest value from
+/// the given producer.
+///
+/// Signals started by the returned producer will not send a value until both
+/// inputs have sent at least one value each.
+public func combineLatestWith<T, U>(otherSignalProducer: SignalProducer<U>)(producer: SignalProducer<T>) -> SignalProducer<(T, U)> {
+	return producer.lift(combineLatestWith)(otherSignalProducer)
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: How freaking awesome is this?

**Comment by jspahrsummers** [comment]
Trying to implement everything here was a mistake, since we're basically talking about the whole framework. So, instead, I've annotated outstanding TODOs and would like to merge this as-is (and create smaller PRs into `swift-development` after that).

:raised_hands:

**Comment by jspahrsummers** [comment]
Gonna #yolo this so I can create dependencies that don't have crazy diffs. :trollface:

**Pull request merged**